From: Kerin Millar <kerframil@gmail.com>

Disable the use of pmi in the vesafb framebuffer driver where the kernel 
is non-modular and PAX_KERNEXEC is enabled, thus resolving a compile 
error. Closes bug 197626. 

This patch is present in upstream grsecurity patches as of
pax-linux-2.6.24.2-test24.patch. This patch can be dropped for any
hardened-sources-2.6.24 based upon pax-linux-2.6.24.2-test24.patch or
later.

--- a/drivers/video/vesafb.c	2008-02-14 20:14:52.000000000 +0000
+++ b/drivers/video/vesafb.c	2008-02-17 21:37:44.000000000 +0000
@@ -302,10 +302,10 @@ static int __init vesafb_probe(struct pl
 
 #ifdef __i386__
 
-#ifdef CONFIG_PAX_KERNEXEC
+#if defined(CONFIG_MODULES) && defined(CONFIG_PAX_KERNEXEC)
 	pmi_code = module_alloc_exec(screen_info.vesapm_size);
 	if (!pmi_code)
-#else
+#elif !defined(CONFIG_PAX_KERNEXEC)
 	if (0)
 #endif
 
@@ -323,13 +323,13 @@ static int __init vesafb_probe(struct pl
 	if (ypan || pmi_setpal) {
 		unsigned short *pmi_base;
 
-#ifdef CONFIG_PAX_KERNEXEC
+#if defined(CONFIG_MODULES) && defined(CONFIG_PAX_KERNEXEC)
 		unsigned long cr0;
 #endif
 
 		pmi_base = (unsigned short*)phys_to_virt(((unsigned long)screen_info.vesapm_seg << 4) + screen_info.vesapm_off);
 
-#ifdef CONFIG_PAX_KERNEXEC
+#if defined(CONFIG_MODULES) && defined(CONFIG_PAX_KERNEXEC)
 		pax_open_kernel(cr0);
 		memcpy(pmi_code, pmi_base, screen_info.vesapm_size);
 		pax_close_kernel(cr0);
@@ -340,7 +340,7 @@ static int __init vesafb_probe(struct pl
 		pmi_start = (void*)((char*)pmi_code + pmi_base[1]);
 		pmi_pal   = (void*)((char*)pmi_code + pmi_base[2]);
 
-#ifdef CONFIG_PAX_KERNEXEC
+#if defined(CONFIG_MODULES) && defined(CONFIG_PAX_KERNEXEC)
 		pmi_start -= __KERNEL_TEXT_OFFSET;
 		pmi_pal -= __KERNEL_TEXT_OFFSET;
 #endif
@@ -487,7 +487,7 @@ static int __init vesafb_probe(struct pl
 	return 0;
 err:
 
-#ifdef CONFIG_PAX_KERNEXEC
+#if defined(CONFIG_MODULES) && defined(CONFIG_PAX_KERNEXEC)
 	module_free_exec(NULL, pmi_code);
 #endif
 
